resolver/resolver.cc

```diff
@@ -365,71 +425,126 @@ class ResolveConstantsWalk {
+                ENFORCE(cursor->scope.isClassOrModule());
+                return cursor->scope.asClassOrModuleRef();
             }
         }
 
-        auto symbol = ctx.state.enterClassSymbol(core::Loc{ctx.file, out->loc}, owner,
-                                                 ast::cast_tree<ast::UnresolvedConstantLit>(out->original)->cnst);
+        return core::ClassOrModuleRef{};
+    }
 
-        auto data = symbol.data(ctx);
-        // force a singleton into existence
-        auto singletonClass = data->singletonClass(ctx);
-        if (possibleGenericType) {
-            ensureTGenericMixin(ctx, singletonClass);
+    // Determine if a child class shares a name in suffix, potentially rooted in any of the nesting scopes.
+    static core::ClassOrModuleRef childPackageOwner(core::MutableContext ctx, const ImportStubs &importStubs,
+                                                    const Nesting *scope, std::vector<ast::ConstantLit *> suffix) {
+        std::vector<core::NameRef> prefix;
+        prefix.reserve(5);
+
+        for (auto *cursor = scope; cursor != nullptr; cursor = cursor->parent.get()) {
+            // If this is expensive, we could cache it in stubForRbiGeneration below
+            if (!scopeToNames(ctx, cursor->scope, prefix)) {
+                continue;
+            }
+
+            if (importStubs.fromChildNamespace(ctx, prefix, suffix)) {
+                ENFORCE(cursor->scope.isClassOrModule());
+                return cursor->scope.asClassOrModuleRef();
+            }
         }
 
-        out->symbol = symbol;
+        return core::ClassOrModuleRef{};
+    }
+
+    static void stubForRbiGeneration(core::MutableContext ctx, const ImportStubs &importStubs, const Nesting *scope,
+                                     ast::ConstantLit *out, bool possibleGenericType) {
+        std::vector<ast::ConstantLit *> suffix;
+        {
+            auto *cursor = out;
+            while (cursor != nullptr) {
+                suffix.emplace_back(cursor);
+                auto &original = ast::cast_tree_nonnull<ast::UnresolvedConstantLit>(cursor->original);
+                cursor = ast::cast_tree<ast::ConstantLit>(original.scope);
+            }
+            absl::c_reverse(suffix);
+        }
+
+        auto owner = core::Symbols::root();
+
+        // First, determine if we're already in the context of a parent package by crawling up the nesting scopes.
+        auto parentPackage = parentPackageOwner(ctx, importStubs, scope, suffix.front());
+        if (parentPackage.exists()) {
+            owner = parentPackage;
+        }
+
+        // If we're not in a parent package, check the name suffix to determine if we're inside of a child package.
+        auto childPackage = childPackageOwner(ctx, importStubs, scope, suffix);
+        if (childPackage.exists()) {
+            owner = childPackage;
+        }
+
+        // If both of the previous two checks fail, stub the constant as defined at the root scope.
```

Comment:
I am 50/50 on whether to retain this comment: it's not wrong, per se, but the code it's describing is already using `owner` as a description of where we're stubbing the constant.  Maybe a slight rework?

Reply:
I rephrased it on the initializer for `owner`, which feels like it sets expectations a bit better for the remainder of the function.